[wip] charmpp : fix build prefix, introduce new property charmarch

s-sajid-ali opened the pull request:
fixes #15803

Rationale : As explained in https://github.com/spack/spack/pull/15186#issuecomment-591701729

I had to add `-lpthread` as without this I was getting errors regarding undefined symbols. I temporarily commented out the `backend=mpi provides` as this was causing concretization errors (when building with `+openmpi/pmix` as specified in `namd` instructions). What would be the long term fix for this ? 

Successfully built spec documented [here](https://pastebin.com/N6X365Xa).

Sanity check : 
```
[sajid@xrmlite ~]$ cd $(spack location -i /5pq)/bin
[sajid@xrmlite bin]$ ./charmrun ./namd2

Running on 1 processors:  ./namd2
charmrun>  /usr/bin/setarch x86_64 -R  mpirun -np 1  ./namd2
This version of Spack (openmpi ~legacylaunchers schedulers=slurm)
is installed without the mpiexec/mpirun commands to prevent
unintended performance issues. See https://github.com/spack/spack/pull/10340
for more details.
^C
```

Requesting @matthiasdiener for review !

matthiasdiener commented on var/spack/repos/builtin/packages/charmpp/package.py:
```diff
@@ -56,10 +56,18 @@ class Charmpp(Package):
+    variant(
+        "ucx-pmi",
+        default="simplePMI",
+        values=("simplePMI", "slrumPMI", "slurmPMI2", "PMIx"),
```
Spelling error:  `"slrumPMI"`

matthiasdiener commented:
Looks good imo. I'm not sure about the namd/pthreads issue, but this might be better in a separate PR? Regarding the `backend=mpi provides`, I think this is not critical functionality and can be commented out for now. I'm not sure what the long-term fix for this is.

s-sajid-ali commented:
Any update on https://github.com/spack/spack/pull/15186#issuecomment-590527904 ? 

I'll fix the flake8 issues soon.

matthiasdiener commented on var/spack/repos/builtin/packages/charmpp/package.py:
```diff
@@ -18,10 +18,12 @@ class Charmpp(Package):
-    version("develop", branch="charm")
+    version("develop", branch="master")
+
+    version('6.10.1', sha256='ab96198105daabbb8c8bdf370f87b0523521ce502c656cb6cd5b89f69a2c70a8')
```
Now that #15667 is merged to `develop`, these arent' necessary anymore.

s-sajid-ali replied:
This was causing a merge conflict and preventing the CI from running so I manually copied the contents of https://github.com/spack/spack/pull/15667 over here. I guess I could've added that commit on my branch but it was simple enough to cut/paste.

matthiasdiener replied:
I think this might cause conflicts when merging later on. I think it might be easier to just run `git merge develop` on your branch, fix the conflict, and push it.

s-sajid-ali replied:
done!

matthiasdiener commented on var/spack/repos/builtin/packages/charmpp/package.py:
```diff
@@ -56,10 +58,18 @@ class Charmpp(Package):
+    # https://github.com/UIUC-PPL/charm/issues/2477
+    variant(
+        "ucx-pmi",
+        default="simplePMI",
```
As far as I know, PMIx is the version with the best support in Charm++ currently, maybe this could be the default.

s-sajid-ali replied:
For both ucx and non-ucx installs ?

matthiasdiener replied:
I think PMI is only used when building with ucx, but @nitbhat may be able to confirm.

nitbhat replied:
All of gni, ofi and ucx use PMI.

matthiasdiener commented:
Also pinging @nitbhat for review.

s-sajid-ali commented:
Looks like the `+pthreads` variant is not supported anymore in the latest release. When was it deprecated ?

matthiasdiener commented:
> Looks like the `+pthreads` variant is not supported anymore in the latest release. When was it deprecated ?

As far as I know, it is still supported. Do you get an error when building with it?

s-sajid-ali commented:
When I build without `pthreads`, `charmpp` still embeds `pthreads` symbols that cause `namd` build to fail. 
```
[sajid@xrm-backup ~]$ spack find -lv charmpp
==> 1 installed package
-- linux-rhel8-skylake_avx512 / gcc@7.5.0 -----------------------
lhi4f4e charmpp@6.10.1 backend=ucx build-target=LIBS ~cuda~omp~papi+production~pthreads+shared+smp~syncft~tcp~tracing ucx-pmi=PMIx
[sajid@xrm-backup ~]$ cd $(spack location -i /lhi)
[sajid@xrm-backup charmpp-6.10.1-lhi4f4ek4kgyd5bj56di2joldyaoldxa]$ nm ucx-linux-x86_64/lib/libmoduleC
libmoduleCkCache.a                    libmoduleCkLoop.a                     libmoduleComboCentLB.a                libmoduleCommonLBs.a
libmoduleCkIO.a                       libmoduleCkMulticast.a                libmoduleCommAwareRefineLB.a
libmoduleCkIO.dep                     libmoduleCkSparseContiguousReducer.a  libmoduleCommLB.a
[sajid@xrm-backup charmpp-6.10.1-lhi4f4ek4kgyd5bj56di2joldyaoldxa]$ nm ucx-linux-x86_64/lib/libmoduleCkLoop.a | grep pthread
                 U pthread_attr_init
                 U pthread_attr_setdetachstate
                 U pthread_cond_signal
                 U pthread_create
                 U pthread_join
                 U pthread_mutex_lock
                 U pthread_mutex_unlock
[sajid@xrm-backup charmpp-6.10.1-lhi4f4ek4kgyd5bj56di2joldyaoldxa]$
```

And I'm unable to enable the `pthreads` variant : 
```
[sajid@xrm-backup ~]$ spack find -lv openmpi
==> 1 installed package
-- linux-rhel8-skylake_avx512 / gcc@7.5.0 -----------------------
dkstt2g openmpi@4.0.2~cuda+cxx_exceptions fabrics=ucx ~java~legacylaunchers~memchecker patches=073477a76bba780c67c36e959cd3ee6910743e2735c7e76850ffba6791d498e4 +pmi schedulers=slurm ~sqlite3~thread_multiple+vt
[sajid@xrm-backup ~]$ spack install charmpp@6.10.1+pthreads backend=ucx ucx-pmi=PMIx ^/dks
.....
==> 16181: Installing charmpp
==> Using cached archive: /home/sajid/packages/spack/var/spack/cache/_source-cache/archive/ab/ab96198105daabbb8c8bdf370f87b0523521ce502c656cb6cd5b89f69a2c70a8.tar.gz
==> Staging archive: /tmp/sajid/spack-stage/spack-stage-charmpp-6.10.1-hij6dby3nhfc5hyx4dmtf2jui42qd5bs/charm-6.10.1.tar.gz
==> Created stage in /tmp/sajid/spack-stage/spack-stage-charmpp-6.10.1-hij6dby3nhfc5hyx4dmtf2jui42qd5bs
==> No patches needed for charmpp
==> 16181: charmpp: Building charmpp [Package]
==> 16181: charmpp: Executing phase: 'install'
==> Error: ProcessError: Command exited with status 1:
    './build' 'LIBS' 'ucx-linux-x86_64' 'gcc' 'gfortran' '-j16' '--destination=/home/sajid/packages/spack/opt/spack/linux-rhel8-skylake_avx512/gcc-7.5.0/charmpp-6.10.1
-hij6dby3nhfc5hyx4dmtf2jui42qd5bs/ucx-linux-x86_64' '--basedir=/home/sajid/packages/spack/opt/spack/linux-rhel8-skylake_avx512/gcc-7.5.0/ucx-1.6.1-3bbygc4yx7l5tm54oioy
wbqkeracssou' 'smp' 'pthreads' '--build-shared' '--with-production'
See build log for details:
  /tmp/sajid/spack-stage/spack-stage-charmpp-6.10.1-hij6dby3nhfc5hyx4dmtf2jui42qd5bs/spack-build-out.txt

==> Error: Failed to install charmpp due to ChildError: ProcessError: Command exited with status 1:
    './build' 'LIBS' 'ucx-linux-x86_64' 'gcc' 'gfortran' '-j16' '--destination=/home/sajid/packages/spack/opt/spack/linux-rhel8-skylake_avx512/gcc-7.5.0/charmpp-6.10.1
-hij6dby3nhfc5hyx4dmtf2jui42qd5bs/ucx-linux-x86_64' '--basedir=/home/sajid/packages/spack/opt/spack/linux-rhel8-skylake_avx512/gcc-7.5.0/ucx-1.6.1-3bbygc4yx7l5tm54oioy
wbqkeracssou' 'smp' 'pthreads' '--build-shared' '--with-production'See build log for details:
  /tmp/sajid/spack-stage/spack-stage-charmpp-6.10.1-hij6dby3nhfc5hyx4dmtf2jui42qd5bs/spack-build-out.txt
Traceback (most recent call last):
  File "/home/sajid/packages/spack/lib/spack/spack/build_environment.py", line 801, in child_process
    return_value = function()
  File "/home/sajid/packages/spack/lib/spack/spack/installer.py", line 1050, in build_process
    phase(pkg.spec, pkg.prefix)
  File "/home/sajid/packages/spack/lib/spack/spack/package.py", line 108, in phase_wrapper
    phase(spec, prefix)
  File "/home/sajid/packages/spack/var/spack/repos/builtin/packages/charmpp/package.py", line 249, in install
  File "/home/sajid/packages/spack/lib/spack/spack/util/executable.py", line 189, in __call__
    proc.returncode, long_msg)
spack.util.executable.ProcessError: Command exited with status 1:
    './build' 'LIBS' 'ucx-linux-x86_64' 'gcc' 'gfortran' '-j16' '--destination=/home/sajid/packages/spack/opt/spack/linux-rhel8-skylake_avx512/gcc-7.5.0/charmpp-6.10.1
-hij6dby3nhfc5hyx4dmtf2jui42qd5bs/ucx-linux-x86_64' '--basedir=/home/sajid/packages/spack/opt/spack/linux-rhel8-skylake_avx512/gcc-7.5.0/ucx-1.6.1-3bbygc4yx7l5tm54oioy
wbqkeracssou' 'smp' 'pthreads' '--build-shared' '--with-production'
```

Cause of build failure : 
```
[sajid@xrm-backup ~]$ cat   /tmp/sajid/spack-stage/spack-stage-charmpp-6.10.1-hij6dby3nhfc5hyx4dmtf2jui42qd5bs/spack-build-out.txt
==> 16181: charmpp: Executing phase: 'install'
==> [2020-03-25-12:04:29.683222, 16193] './build' 'LIBS' 'ucx-linux-x86_64' 'gcc' 'gfortran' '-j16' '--destination=/home/sajid/packages/spack/opt/spack/linux-rhel8-sky
lake_avx512/gcc-7.5.0/charmpp-6.10.1-hij6dby3nhfc5hyx4dmtf2jui42qd5bs/ucx-linux-x86_64' '--basedir=/home/sajid/packages/spack/opt/spack/linux-rhel8-skylake_avx512/gcc-
7.5.0/ucx-1.6.1-3bbygc4yx7l5tm54oioywbqkeracssou' 'smp' 'pthreads' '--build-shared' '--with-production'
Error> option: pthreads is not supported in this version!
Supported compilers: clang craycc gcc icc iccstatic msvc pgcc xlc xlc64
Supported options: bigemulator bigsim common cuda flang gfortran ifort nolb omp ompipmix ooc papi pedantic perftools persistent pgf90 slurmpmi slurmpmi2 smp tsan
```

s-sajid-ali commented:
I overcame this issue by building `charmpp~pthreads` and explicitly passing `-lpthreads` as a compiler flag to `namd`.  Is there a better fix for the same ?

matthiasdiener commented:
> When I build without `pthreads`, `charmpp` still embeds `pthreads` symbols that cause `namd` build to fail.
> 
> ```
> [sajid@xrm-backup ~]$ spack find -lv charmpp
> ==> 1 installed package
> -- linux-rhel8-skylake_avx512 / gcc@7.5.0 -----------------------
> lhi4f4e charmpp@6.10.1 backend=ucx build-target=LIBS ~cuda~omp~papi+production~pthreads+shared+smp~syncft~tcp~tracing ucx-pmi=PMIx
> [sajid@xrm-backup ~]$ cd $(spack location -i /lhi)
> [sajid@xrm-backup charmpp-6.10.1-lhi4f4ek4kgyd5bj56di2joldyaoldxa]$ nm ucx-linux-x86_64/lib/libmoduleC
> libmoduleCkCache.a                    libmoduleCkLoop.a                     libmoduleComboCentLB.a                libmoduleCommonLBs.a
> libmoduleCkIO.a                       libmoduleCkMulticast.a                libmoduleCommAwareRefineLB.a
> libmoduleCkIO.dep                     libmoduleCkSparseContiguousReducer.a  libmoduleCommLB.a
> [sajid@xrm-backup charmpp-6.10.1-lhi4f4ek4kgyd5bj56di2joldyaoldxa]$ nm ucx-linux-x86_64/lib/libmoduleCkLoop.a | grep pthread
>                  U pthread_attr_init
>                  U pthread_attr_setdetachstate
>                  U pthread_cond_signal
>                  U pthread_create
>                  U pthread_join
>                  U pthread_mutex_lock
>                  U pthread_mutex_unlock
> [sajid@xrm-backup charmpp-6.10.1-lhi4f4ek4kgyd5bj56di2joldyaoldxa]$
> ```

I think that is expected. CkLoop uses pthreads internally.

> 
> And I'm unable to enable the `pthreads` variant :
> 
> ```
> [sajid@xrm-backup ~]$ spack find -lv openmpi
> ==> 1 installed package
> -- linux-rhel8-skylake_avx512 / gcc@7.5.0 -----------------------
> dkstt2g openmpi@4.0.2~cuda+cxx_exceptions fabrics=ucx ~java~legacylaunchers~memchecker patches=073477a76bba780c67c36e959cd3ee6910743e2735c7e76850ffba6791d498e4 +pmi schedulers=slurm ~sqlite3~thread_multiple+vt
> [sajid@xrm-backup ~]$ spack install charmpp@6.10.1+pthreads backend=ucx ucx-pmi=PMIx ^/dks
> .....
> ==> 16181: Installing charmpp
> ==> Using cached archive: /home/sajid/packages/spack/var/spack/cache/_source-cache/archive/ab/ab96198105daabbb8c8bdf370f87b0523521ce502c656cb6cd5b89f69a2c70a8.tar.gz
> ==> Staging archive: /tmp/sajid/spack-stage/spack-stage-charmpp-6.10.1-hij6dby3nhfc5hyx4dmtf2jui42qd5bs/charm-6.10.1.tar.gz
> ==> Created stage in /tmp/sajid/spack-stage/spack-stage-charmpp-6.10.1-hij6dby3nhfc5hyx4dmtf2jui42qd5bs
> ==> No patches needed for charmpp
> ==> 16181: charmpp: Building charmpp [Package]
> ==> 16181: charmpp: Executing phase: 'install'
> ==> Error: ProcessError: Command exited with status 1:
>     './build' 'LIBS' 'ucx-linux-x86_64' 'gcc' 'gfortran' '-j16' '--destination=/home/sajid/packages/spack/opt/spack/linux-rhel8-skylake_avx512/gcc-7.5.0/charmpp-6.10.1
> -hij6dby3nhfc5hyx4dmtf2jui42qd5bs/ucx-linux-x86_64' '--basedir=/home/sajid/packages/spack/opt/spack/linux-rhel8-skylake_avx512/gcc-7.5.0/ucx-1.6.1-3bbygc4yx7l5tm54oioy
> wbqkeracssou' 'smp' 'pthreads' '--build-shared' '--with-production'
> See build log for details:
>   /tmp/sajid/spack-stage/spack-stage-charmpp-6.10.1-hij6dby3nhfc5hyx4dmtf2jui42qd5bs/spack-build-out.txt
> 
> ==> Error: Failed to install charmpp due to ChildError: ProcessError: Command exited with status 1:
>     './build' 'LIBS' 'ucx-linux-x86_64' 'gcc' 'gfortran' '-j16' '--destination=/home/sajid/packages/spack/opt/spack/linux-rhel8-skylake_avx512/gcc-7.5.0/charmpp-6.10.1
> -hij6dby3nhfc5hyx4dmtf2jui42qd5bs/ucx-linux-x86_64' '--basedir=/home/sajid/packages/spack/opt/spack/linux-rhel8-skylake_avx512/gcc-7.5.0/ucx-1.6.1-3bbygc4yx7l5tm54oioy
> wbqkeracssou' 'smp' 'pthreads' '--build-shared' '--with-production'See build log for details:
>   /tmp/sajid/spack-stage/spack-stage-charmpp-6.10.1-hij6dby3nhfc5hyx4dmtf2jui42qd5bs/spack-build-out.txt
> Traceback (most recent call last):
>   File "/home/sajid/packages/spack/lib/spack/spack/build_environment.py", line 801, in child_process
>     return_value = function()
>   File "/home/sajid/packages/spack/lib/spack/spack/installer.py", line 1050, in build_process
>     phase(pkg.spec, pkg.prefix)
>   File "/home/sajid/packages/spack/lib/spack/spack/package.py", line 108, in phase_wrapper
>     phase(spec, prefix)
>   File "/home/sajid/packages/spack/var/spack/repos/builtin/packages/charmpp/package.py", line 249, in install
>   File "/home/sajid/packages/spack/lib/spack/spack/util/executable.py", line 189, in __call__
>     proc.returncode, long_msg)
> spack.util.executable.ProcessError: Command exited with status 1:
>     './build' 'LIBS' 'ucx-linux-x86_64' 'gcc' 'gfortran' '-j16' '--destination=/home/sajid/packages/spack/opt/spack/linux-rhel8-skylake_avx512/gcc-7.5.0/charmpp-6.10.1
> -hij6dby3nhfc5hyx4dmtf2jui42qd5bs/ucx-linux-x86_64' '--basedir=/home/sajid/packages/spack/opt/spack/linux-rhel8-skylake_avx512/gcc-7.5.0/ucx-1.6.1-3bbygc4yx7l5tm54oioy
> wbqkeracssou' 'smp' 'pthreads' '--build-shared' '--with-production'
> ```
> 
> Cause of build failure :
> 
> ```
> [sajid@xrm-backup ~]$ cat   /tmp/sajid/spack-stage/spack-stage-charmpp-6.10.1-hij6dby3nhfc5hyx4dmtf2jui42qd5bs/spack-build-out.txt
> ==> 16181: charmpp: Executing phase: 'install'
> ==> [2020-03-25-12:04:29.683222, 16193] './build' 'LIBS' 'ucx-linux-x86_64' 'gcc' 'gfortran' '-j16' '--destination=/home/sajid/packages/spack/opt/spack/linux-rhel8-sky
> lake_avx512/gcc-7.5.0/charmpp-6.10.1-hij6dby3nhfc5hyx4dmtf2jui42qd5bs/ucx-linux-x86_64' '--basedir=/home/sajid/packages/spack/opt/spack/linux-rhel8-skylake_avx512/gcc-
> 7.5.0/ucx-1.6.1-3bbygc4yx7l5tm54oioywbqkeracssou' 'smp' 'pthreads' '--build-shared' '--with-production'
> Error> option: pthreads is not supported in this version!
> Supported compilers: clang craycc gcc icc iccstatic msvc pgcc xlc xlc64
> Supported options: bigemulator bigsim common cuda flang gfortran ifort nolb omp ompipmix ooc papi pedantic perftools persistent pgf90 slurmpmi slurmpmi2 smp tsan
> ```

The `pthreads` option is only available in netlrts and mpi.

matthiasdiener commented:
> I overcame this issue by building `charmpp~pthreads` and explicitly passing `-lpthreads` as a compiler flag to `namd`. Is there a better fix for the same ?

I think we might want to add `-lpthread` (or better `-pthread`) in charm itself when building ucx without smp (when building ucx *with* smp, that flag should already be there). For now, I think your workaround is best.

s-sajid-ali commented:
>The pthreads option is only available on the Netlrts and MPI network layers.

How do I add a conflict for two conditions ? 
Alternatively we could do :
```
if ("backend!=mpi" in self.spec) or ("backend!=netlrts" in self.spec):
    if `+pthreads` in self.spec:
        raiseError
```

> I think we might want to add -lpthread (or better -pthread) in charm itself when building ucx without smp (when building ucx with smp, that flag should already be there). 

Would this be a makefile patch ?

matthiasdiener commented:
> > The pthreads option is only available on the Netlrts and MPI network layers.
> 
> How do I add a conflict for two conditions ?
> Alternatively we could do :
> 
> ```
> if ("backend!=mpi" in self.spec) or ("backend!=netlrts" in self.spec):
>     if `+pthreads` in self.spec:
>         raiseError
> ```

Maybe just add a conflict for all non-supported cases? Something like:

```
conflicts("backend=multicore", "+pthreads")
....
```

> 
> > I think we might want to add -lpthread (or better -pthread) in charm itself when building ucx without smp (when building ucx with smp, that flag should already be there).
> 
> Would this be a makefile patch ?

It would be a patch in our build system, something like https://github.com/UIUC-PPL/charm/pull/2778 . Please give this a try if you have time.

s-sajid-ali commented:
Does everything look good now ?

matthiasdiener commented:
> > > The pthreads option is only available on the Netlrts and MPI network layers.
> 
> > 
> 
> > How do I
> 
> > > I think we might want to add -lpthread (or better -pthread) in charm itself when building ucx without smp (when building ucx with smp, that flag should already be there).
> 
> > 
> 
> > Would this be a makefile patch ?
> 
> 
> 
> It would be a patch in our build system, something like https://github.com/UIUC-PPL/charm/pull/2778 . Please give this a try if you have time.

The patch above has been merged. Can you check if this issue with NAMD still exists?

s-sajid-ali commented:
New error when building `namd ^charmpp@develop` : `https://pastebin.com/MjJf9Uq7`

matthiasdiener commented:
> New error when building `namd ^charmpp@develop` : https://pastebin.com/MjJf9Uq7

Ok, this looks like a separate error from the pthreads issue. By the way, you can attach files directly here on the github issue. 

The patch looks good to me.

scheibelp commented:
Closing/reopening to restart tests

scheibelp commented:
I see what's going wrong, it's not the fault of this PR: this PR removes `charmpp` as an MPI provider, and one of our unit tests is getting hung up on an a different MPI provider. I'll be working on a PR to address that in the next couple days (and that should make this PR pass).

s-sajid-ali commented:
While I added an option to configure `charmpp` with `pmix`, this can't recipe can't build it since the spack recipe for `openmpi` does not have a `pmix` variant (yet!). 

The `charmpp` [manual](https://github.com/UIUC-PPL/charm/blob/master/doc/charm%2B%2B/manual.rst) states that "in order to use the other supported process management interfaces, it is required to have a non-OpenMPI based MPI implementation installed on the system (e.g. Intel MPI, MVAPICH, MPICH, etc.)". 

If spack users install `namd+charmpp` then the mpi implementation gets loaded into the environment due to `fftw+mpi` dependency. Would it be a good idea to add a run only `mpi` dependency to `charmpp` itself (and add a conflict statement depending on pmi variant) ?

I changed all variant names to lowercase as that's more common in other packages.

scheibelp commented:
> If spack users install namd+charmpp then the mpi implementation gets loaded into the environment due to fftw+mpi dependency. Would it be a good idea to add a run only mpi dependency to charmpp itself (and add a conflict statement depending on pmi variant) ?

I think this is saying that `namd` needs to load `openmpi` when `charmpp` uses `openmpi` - is that correct? Adding `openmpi` as a run dependency could be useful to ensure that the environment includes what you want, but IMO we should probably offer more flexibility in what packages are loaded into the environment (i.e. transitive link dependencies in some cases). An interim fix (if you are willing to wait for that) could be to explicitly load `openmpi` in the environment that uses `namd` with `charmpp`.

Side note: I merged #16003 but had to revert it because of other test interactions that have cropped up since that PR was created. There is a WIP PR at https://github.com/spack/spack/pull/16429. I anticipate finishing it by 5/5

s-sajid-ali commented:
> I think this is saying that namd needs to load openmpi when charmpp uses openmpi - is that correct? 

The exact quote from the `charm++` manual states : 
```
After installing UCX, there are several supported process management interfaces (PMI) that
can be specified as options in order to build Charm++ with UCX. These include Simple PMI,
Slurm PMI, Slurm PMI 2 and PMIx (using OpenMPI). Currently, in order to use PMIx for
process management, it is required to have OpenMPI installed on the system. Additionally, 
in order to use the other supported process management interfaces, it is required to have 
a non-OpenMPI based MPI implementation installed on the system (e.g. Intel MPI, MVAPICH, 
MPICH, etc.).
```
Thus, it's a (runtime?) dependency of `charm++` itself. I added `mpi` as a dependency for each `pmi` variant individually as well as a conflict for `pmi!=pmix` and `+openmpi`.  As stated earlier, the `pmi=pmix` case is provided to those who have external installs of `openmpi+pmix`. This should load the correct mpi into the runtime environment for `charm++` and any dependent packages as well.

I also added the git repo for `namd` [as documented](https://www.ks.uiuc.edu/Research/namd/development.html), but this fails on my workstation running `centos-8` (though it works on my laptop running `wsl2-ubuntu-18.04` ) with :
```
[sajid@xrmlite temp]$ git clone https://charm.cs.illinois.edu/gerrit/namd.git
Cloning into 'namd'...
fatal: unable to access 'https://charm.cs.illinois.edu/gerrit/namd.git/': error:1425F102:SSL 
routines:ssl_choose_client_version:unsupported protocol
```
Should I keep the git url for now and let the fetch options be fixed later or remove it ?

Thanks for cleaning up the mpi tests! Once https://github.com/spack/spack/pull/16429 is merged, the CI for this can be re-run.

s-sajid-ali commented:
@scheibelp : Any clue as to why the tests detect `charmpp` as an mpi provider even though this PR disables that ?

scheibelp commented:
> Any clue as to why the tests detect charmpp as an mpi provider even though this PR disables that ?

Let me try closing/reopening this - I'm trying to remember when the CI decides to rebase on develop and I think closing/reopening is one of the cases where it does.

s-sajid-ali commented:
Looks like that didn't work, would merging develop into this branch help ?

scheibelp commented:
> would merging develop into this branch help ?

Yes that would work for sure - thanks!

scheibelp commented:
@alalazo just pointed out to me that there's a test that actually needs to be updated here: `test_provider_lists` is trying to assert that `charmpp` is an MPI provider (i.e. there was an additional problem to the one resolved by #16429), but this PR has removed it. Apologies to all for missing that.

@s-sajid-ali would you be able to update that unit test?

s-sajid-ali commented:
@scheibelp : That fixed the issue! Could this PR get a final review now that the tests pass?

scheibelp commented:
LGTM - thanks!
